### ISSUE TITLE
Pointers / Teleport non-functional yet

### ISSUE THREAD

**Issue by SimonDarksideJ** [Innoactive/IA-unity-VR-toolkit-VRTK#15]
Possibly related to the controller issue.

Loading up any of the Pointer/Teleport scenes doesn't work.  

In fact the only scene using the controller I got to work was the one with the RC car (scene 15) which used the right thumbstick (left didn't work?).
However, the scene is titled "TouchPadAxisControl" which could indicate it's using the wrong control type anyway (thumbstick instead of touchpad?)

**Comment by tomwim** [comment]
That is interesting because teleporting works perfectly fine on my side. Never had any issues with that. Currently there are only 6 example scenes I couldn't get to work properly (12, 26, 28, 35, 37, 39).

I'll include the changes from the other issue (Controller update) and let you know again.

**Comment by tomwim** [comment]
I included the changes and tested on 2 PCs. The controller works for me on both, in Editor as well as built.

**Comment by SimonDarksideJ** [comment]
OK, I'll grab the latest update and test with the new MRTP 4 client and see if that's improved.
The controllers didn't even show up in the editor for myself.

**Comment by tomwim** [comment]
That is something I have a problem with as well. Maybe you have a solution for it.
Before I click play in Unity I have to close the Windows Mixed Reality Portal, otherwise the controllers are not tracked. I also need to first exit Play Mode before closing the WinMR Portal or Unity freezes...
Do you have the same problem and a way around that?

**Comment by SimonDarksideJ** [comment]
OK, interesting results from those scenes (personally, the scenes do need some "demo" text" to tell you what the demo is for in VR :D)

* 12 - looks good, can see both hands, can't move
* 26 - Only right hand visible but can't interact. Left hand not visible but can knock over towers. Can't pickup
* 28 - Both hands show, can interact with door with grab button. Can't move with teleport.  Doot "MOVES" as you walk around it??? which causes sickness ;S
* 35 - Most interesting Scene, can't interact but both controllers show.  ALL buttons work with the exception of the thumbstick, *which might explain why teleporting isn't working.*  
* 37 - RIGHT, so Teleporting worked using the TOUCHPADS in this scene, which was good.  However, it demonstrates something wrong either with MRTK or VRTK with gravity or floor positioning.  As I climbed up half a wall then was stuck at that height when I teleported??
* 39 - Freakishly bad, I tried to teleport and it broke the scene and nearly crashed Unity.

So to sum up

[ ] Thumbsticks not being recognised on controllers
[ ] Floor positioning is off. Need to see if it's MR, Unity or VRTK (this was a big issue on MRTK)
[ ] Teleporting and relative height to teleport target, again MR, Unity or VRTK?
[ ] The thumbstick press sometimes got "lost" which resulted in a teleport.  I suspect a side affect of using the update over events.

> interesting side note, each scene when opened, asked to save on exit. Possibly the MRTP 4 build is changing scene yaml?

Great progress

**Comment by tomwim** [comment]
Thanks for the testing!

26 - Im currently working on this. Both swords are attached to the left controller, and that causes them to be not visible (?). If you disable one, the other is visible. Has something to do with wrong indices I assume. Probably something wrong in my MR_Controller script.
28 - Totally agree here. Couldn't even find out what the scene is for to be honest :D
35 - Same behavior. The controllers show the tooltips and get transparent but no interaction possible.
37 - I think the rigidbody is not attached to the correct object. Should be an issue on VRTK/my side here. That's why you're not falling and are stuck at a certain height after climbing up and then teleport (relative to the floor you teleported to).
39 - gives me constant XR Error with Texture Array. Leads to no picture in the HMD.

Additional: 12 - The middle point of the area is not where it belongs and gets more and more off after each teleport. There is an offset added to it that if removed solves the problem, but that is logic in the original scripts.

Thumbsticks - the input from the controllers is stored but there is currently no bridge between Thumbstick and VRTK if I'm not wrong. Since Oculus has only the Thumbstick and Vive has only the Touchpad they are treated the same. Oculus Thumbstick is considered a Touchpad in VRTK. WinMR has both so there is currently logic in the basic Controller of VRTK missing.

Floor position - As I said, seems like an issue on my side still (?). But if MRTK had similar issues, maybe that is the reason...

Relative Height - Either Rigidbody or related to floor position?

Thumbstick press lost - The Thumbstick is handled in events now. Only the Select button is in Update since that caused massive problems on all ends.

**Comment by SimonDarksideJ** [comment]
> Thumbstick press lost - The Thumbstick is handled in events now. Only the Select button is in Update since that caused massive problems on all ends.
Lol, do you mean Thumbstick which isn't working or trackpad :D 

If VRTK only has one medium, then I would map it to the Thumbstick, as that is the default behaviour for MR.  Touchpad would then need to be an extension or something users handle either separately in their projects, or possibly work with VRTK on "additional" input methods.

P.S. Try updating to the newer MRTP4 build, should resolve your Unity error bugs.

**Comment by tomwim** [comment]
Thanks for the MRTP4 tip! That solved some problems.

Update:
26 - Fixed this issue by assigning default indices to each controller based on their handedness.
39 - Works now (thanks to MRTP4). I think in the scene there is no teleporting possible but clicking the Touchpad adds massive amounts of lightning balls. This might have caused the crash?

Thumbsticks - Since teleporting within VRTK is implemented differently than in the MR examples and it feels really weird with the Thumbstick, I added a bool flag to the controller (WindowsMR_TrackedObject) to decide whether the Thumbstick should override the Touchpad. This will be handled differently in the future (see https://github.com/thestonefox/VRTK/issues/1594).

**Comment by tomwim** [comment]
28 - Also works for me now. Probably issues with Thumbstick/Touchpad. Just feels super weird to actually use.

I opened issues for the remaining problems so not everything is in here. So I close this one.
Thanks for all the help!
